flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/UnionResultSubpartitionView.java

```diff
@@ -0,0 +1,248 @@
+ *       the buffer, these buffers will be output consecutively without the entry of buffers from
+ *       other subpartitions in between.
+ * </ul>
+ */
+public class UnionResultSubpartitionView
+        implements ResultSubpartitionView, BufferAvailabilityListener {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UnionResultSubpartitionView.class);
+
+    /** The maximum number of buffers to be cached in an instance of this class. */
+    private static final int CACHE_CAPACITY = 10;
+
+    private final Object lock = new Object();
+
+    /** All the {@link ResultSubpartitionView}s managed by this class. */
+    private final Map<Integer, ResultSubpartitionView> allViews = new HashMap<>();
+
+    /** All the {@link ResultSubpartitionView}s that have data available. */
+    private final SubpartitionSelector<ResultSubpartitionView> availableViews =
+            new RoundRobinSubpartitionSelector<>();
+
+    private final BufferAvailabilityListener availabilityListener;
+
+    private final Queue<ResultSubpartition.BufferAndBacklog> cachedBuffers = new LinkedList<>();
+
+    private boolean isReleased;
+
+    private int sequenceNumber;
+
+    public UnionResultSubpartitionView(BufferAvailabilityListener availabilityListener) {
+        this.availabilityListener = availabilityListener;
+        this.isReleased = false;
+        this.sequenceNumber = 0;
+    }
+
+    public void notifyViewCreated(int subpartitionIndex, ResultSubpartitionView view) {
+        allViews.put(subpartitionIndex, view);
+    }
+
+    @Nullable
+    @Override
+    public ResultSubpartition.BufferAndBacklog getNextBuffer() throws IOException {
+        synchronized (lock) {
+            cacheBuffer();
+            ResultSubpartition.BufferAndBacklog buffer = cachedBuffers.poll();
+
+            if (buffer == null) {
+                return null;
+            }
+
+            return new ResultSubpartition.BufferAndBacklog(
+                    buffer.buffer(),
+                    cachedBuffers.size(),
+                    cachedBuffers.isEmpty()
+                            ? Buffer.DataType.NONE
+                            : cachedBuffers.peek().buffer().getDataType(),
+                    sequenceNumber++);
+        }
+    }
+
+    private void cacheBuffer() throws IOException {
+        do {
+            ResultSubpartitionView currentView = availableViews.getNextSubpartitionToConsume();
+            if (currentView == null) {
+                break;
+            }
+
+            ResultSubpartition.BufferAndBacklog buffer = currentView.getNextBuffer();
+            if (buffer == null) {
+                availableViews.markLastConsumptionStatus(false, false);
+                if (!availableViews.isMoreSubpartitionSwitchable()) {
+                    break;
+                } else {
+                    continue;
+                }
+            }
+
+            boolean isLastBufferPartialRecord =
```

Comment:
Why do we determine `isLastBufferPartialRecord` according to the buffer data type, we can add some annotations here to describe the logic.